### PR TITLE
Add pause/resume options to the bag recorder

### PR DESCRIPTION
**Pull request by ivanpauno**
Implements half of https://github.com/ros2/rosbag2/issues/558: ability to pause/resume recording.

### CODE REVIEW

File: rosbag2_transport/include/rosbag2_transport/recorder.hpp
Comment (MichaelOrlov) [inline rosbag2_transport/include/rosbag2_transport/recorder.hpp:95]: Typo in `toogle_paused()` => `toggle_paused()`

**Review by MichaelOrlov — DISMISSED** [review]
@ivanpauno Thanks for your contribution.
I think that `paused_` defined as `atomic<int>` is overkill and I don't see a real use case when it could be justified to use. I would propose to change it to the `std::atomic<bool>` or `std::atomic_bool`.

I would hide `toggle_paused()` API as private method unless you need it outside. e.g. in Rqtbag. It's sort of ambiguous and difficult to determine an outcome from it and would be preferable to use `pause` `resume` explicitly outside of the Rosbag2.

I am also worry about recording termination by `CRL+C` keyboard binding. As far as I understand, before start using KeyboardHandler in recording `SIGINT` = `CTRL+C` was handling somewhere inside `Node` class and then re-translating to the `SIGTERM` and we were catching and processing `SIGTERM` in [_transport.cpp](https://github.com/ros2/rosbag2/blob/master/rosbag2_py/src/rosbag2_py/_transport.cpp#L163-L166).
However `KeyboardHandler` override `SIGINT` handler on construction and it looks like now `rclcpp::shutdown();` will not be called and we will abnormally exit from process without properly finishing recording.

File: rosbag2_transport/src/rosbag2_transport/recorder.cpp
Comment (ivanpauno) [reply rosbag2_transport/src/rosbag2_transport/recorder.cpp:161]: > @ivanpauno Please correct me if I am wrong, but I don't see a rational to use fetch_xor(1) and integer type for paused_ flag here.

Yeah agreed, it doesn't make much sense.

**Comment by ivanpauno** [comment]
> I am also worry about recording termination by CRL+C keyboard binding. As far as I understand, before start using KeyboardHandler in recording SIGINT = CTRL+C was handling somewhere inside Node class and then re-translating to the SIGTERM and we were catching and processing SIGTERM in _transport.cpp.
However KeyboardHandler override SIGINT handler on construction and it looks like now rclcpp::shutdown(); will not be called and we will abnormally exit from process without properly finishing recording.

Interesting, I didn't thought about signal handling.
I will take another look to that.

File: rosbag2_transport/include/rosbag2_transport/recorder.hpp
Comment (ivanpauno) [reply rosbag2_transport/include/rosbag2_transport/recorder.hpp]: thanks, fixed in https://github.com/ros2/rosbag2/pull/905/commits/67545366f2cea9100a514abcee498ddaf3b6ed70

File: rosbag2_transport/include/rosbag2_transport/recorder.hpp
Comment (ivanpauno) [reply rosbag2_transport/include/rosbag2_transport/recorder.hpp]: Fixed in https://github.com/ros2/rosbag2/pull/905/commits/67545366f2cea9100a514abcee498ddaf3b6ed70.

File: rosbag2_transport/src/rosbag2_transport/recorder.cpp
Comment (ivanpauno) [reply rosbag2_transport/src/rosbag2_transport/recorder.cpp]: Makes sense, see https://github.com/ros2/rosbag2/pull/905/commits/05e55e3edcff6a17659bcae02a9bc71ebb3fe704.

File: rosbag2_transport/src/rosbag2_transport/recorder.cpp
Comment (ivanpauno) [reply rosbag2_transport/src/rosbag2_transport/recorder.cpp]: Fixed in https://github.com/ros2/rosbag2/pull/905/commits/7287b45f155d49b1d5ff0ec0b52862aff521273c.

**Comment by ivanpauno** [comment]
> I would hide toggle_paused() API as private method unless you need it outside. e.g. in Rqtbag. It's sort of ambiguous and difficult to determine an outcome from it and would be preferable to use pause resume explicitly outside of the Rosbag2.

I think it makes sense to offer the same API that the [Player](https://github.com/ros2/rosbag2/blob/4792327ee6962152fa8f0f6b811e68a76c755d66/rosbag2_transport/include/rosbag2_transport/player.hpp#L107-L108) class uses.

File: rosbag2_transport/include/rosbag2_transport/record_options.hpp
Comment (MichaelOrlov) [reply rosbag2_transport/include/rosbag2_transport/record_options.hpp:46]: Though it doesn't make sense to keep it as a variable inside `RecordOptions` if we are not going to allow for users to change it. To avoid confusion and for the sake of the clarity let's remove it from the options for a while.  
I think it will be no problem to add it if we will decide to add feature for allowing to change pause key binding via `RecordOptions` and command line interface. There are no request for this and I think everybody will be happy if it will be a space bar for pause/resume.

**Comment by ivanpauno** [comment]
> I am also worry about recording termination by CRL+C keyboard binding. As far as I understand, before start using KeyboardHandler in recording SIGINT = CTRL+C was handling somewhere inside Node class and then re-translating to the SIGTERM and we were catching and processing SIGTERM in _transport.cpp.

SIGINT isn't being retranslated to sigterm, the default rclcpp sigint signal handler is calling shutdown() asynchronously.

> However KeyboardHandler override SIGINT handler on construction and it looks like now rclcpp::shutdown(); will not be called and we will abnormally exit from process without properly finishing recording.

Yes, this is an issue.
IMO `KeyboardHandler` should have an option to disable installing the signal handler, or it should be using signal chaining by default.
I don't like signal chaining though.

I will open a PR in the `keyboard_handler` repo and link it here.

**Comment by MichaelOrlov** [comment]
> Yes, this is an issue. IMO `KeyboardHandler` should have an option to disable installing the signal handler, or it should be using signal chaining by default. I don't like signal chaining though.
> 
> I will open a PR in the `keyboard_handler` repo and link it here.

I think chaining signal handler would be a valid workaround. I can help with PR for `KeyboardHandler`

**Comment by MichaelOrlov** [comment]
@ivanpauno Wouldn't you mind to add at least one unit test to the https://github.com/ros2/rosbag2/blob/master/rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp
to cover new `--start-paused` command line option?
Ideally it would be nice to  have one more test for pause/resume via  simulating keyboard press. Similar to how it was done for the [Player class](https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/test/rosbag2_transport/test_keyboard_controls.cpp). I hope it would be possible to create some [mock writer](https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/test/rosbag2_transport/test_record.cpp#L62-L75) to check in test if  `writer_->write(..)` has been called.

**Comment by MichaelOrlov** [comment]
@ivanpauno If you wouldn't mind I would suggest that I will take over issue relating to the signal handler in `KeyboardHandler`  https://github.com/ros-tooling/keyboard_handler/issues/9 and you will focus on unit tests for the `pause/resume` options in bag recorder.
What do you think?

**Comment by ivanpauno** [comment]
> @ivanpauno If you wouldn't mind I would suggest that I will take over issue relating to the signal handler in KeyboardHandler ros-tooling/keyboard_handler#9 and you will focus on unit tests for the pause/resume options in bag recorder.
What do you think?

That sounds fine to me!
I've left a comment in the issue.

File: rosbag2_transport/include/rosbag2_transport/record_options.hpp
Comment (ivanpauno) [reply rosbag2_transport/include/rosbag2_transport/record_options.hpp]: Fixed in https://github.com/ros2/rosbag2/pull/905/commits/1f404b9e822976af4bf4e496c613ef2538f5a01a.

**Comment by ivanpauno** [comment]
> Ideally it would be nice to have one more test for pause/resume via simulating keyboard press. Similar to how it was done for the Player class. I hope it would be possible to create some mock writer to check in test if writer_->write(..) has been called.

Test added in https://github.com/ros2/rosbag2/pull/905/commits/b6c35a67b9a90a1808a9b447366e127cb6ebf56f.

**Review by emersonknapp — APPROVED** [review]
API looks good to me - I like a match with the Player. Sounds like you have the shutdown issue under control

**Comment by MichaelOrlov** [comment]
@emersonknapp @ivanpauno I think we shouldn't merge this PR untill https://github.com/ros-tooling/keyboard_handler/pull/10 will be merged, otherwise recording will be broken.

**Comment by MichaelOrlov** [comment]
> Agreed, I will also have to update this PR to use the new signal handling options there.

You can, but you don't have to. 
The correct `CTRL+C` handling will be correctly processed by default via calling original signal handler from rclcpp context after calling `KeyboardHandler` signal handler.

**Comment by ivanpauno** [comment]
> The correct CTRL+C handling will be correctly processed by default via calling original signal handler from rclcpp context after calling KeyboardHandler signal handler.

That's right, I forgot about that.
It seems that we only need to wait until that PR gets merged and then rerun CI here.

**Comment by MichaelOrlov** [comment]
@ivanpauno I see you have added unit test to cover keyboard press handling for pause/resume during recording. 
However I don't see coverage for the new `--start-paused` command line option.
Could you please add one more test to the 
https://github.com/ros2/rosbag2/blob/master/rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp

**Comment by ivanpauno** [comment]
> Could you please add one more test to the
https://github.com/ros2/rosbag2/blob/master/rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp

I'm having issues when running those tests locally, they always timeout even when using the `master` branch.
Do you have any idea what could be happening?

<details>
<summary>Example output</summary>

[==========] Running 16 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 16 tests from RecordFixture
[ RUN      ] RecordFixture.record_end_to_end_test_with_zstd_file_compression
/home/ivanpauno/ros2_ws/src/ros2/rosbag2/rosbag2_tests/test/rosbag2_tests/test_rosbag2_record_end_to_end.cpp:86: Failure
Value of: pub_manager.wait_for_matched(topic_name)
  Actual: false
Expected: true
Expected find rosbag subscription

</details>

I have added a `ros2 bag record --start-paused` test in https://github.com/ros2/rosbag2/pull/905/commits/b11614f0ae676408c539231ba6f70f4baf51d9a5.
It's at least building locally, but IDK if it's passing.

**Comment by ivanpauno** [comment]
Rpr checker failure seems unrelated.

CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15638/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10318/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13306/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15860/)

**Comment by emersonknapp** [comment]
The `end_to_end` tests are marked `xfail` and therefore nothing is enforcing that they keep passing. I am also seeing a failure when running locally.

**Comment by ivanpauno** [comment]
> Are you building all of ros2 up to rosbag2

Yes, I'm doing that.
I have pulled and rebuilt everything. I might have a random issue in my workspace, I will try a clean rebuild.


> The end_to_end tests are marked xfail and therefore nothing is enforcing that they keep passing.

I have checked the output of the `Test rosbag2 / build_and_test (pull_request)` check in the `Run xfail tests` section, and the test I added is passing.

<details>
<summary>output</summary>

```
1: [ RUN      ] RecordFixture.record_end_to_end_test_start_paused
1: stdin is not a terminal device. Keyboard handling disabled.[INFO] [1637682700.124453840] [rosbag2_recorder]: Press SPACE for pausing/resuming
1: [INFO] [1637682700.127076721] [rosbag2_storage]: Opened database '/tmp/tmp_test_dir_ZUrO9M/record_end_to_end_test_start_paused/record_end_to_end_test_start_paused_0.db3' for READ_WRITE.
1: [INFO] [1637682700.127164920] [rosbag2_recorder]: Listening for topics...
1: [INFO] [1637682700.131385888] [rosbag2_recorder]: Subscribed to topic '/test_topic'
1: [INFO] [1637682700.133184375] [rosbag2_recorder]: All requested topics are subscribed. Stopping discovery...
1: [INFO] [1637682700.687889545] [rclcpp]: signal_handler(signum=2)
1: [INFO] [1637682700.913097269] [rosbag2_storage]: Opened database '/tmp/tmp_test_dir_ZUrO9M/record_end_to_end_test_start_paused/record_end_to_end_test_start_paused_0.db3' for READ_ONLY.
1: [       OK ] RecordFixture.record_end_to_end_test_start_paused (1262 ms)
```

</details>

**Comment by emersonknapp** [comment]
I'm inclined to move forward, given that everything is green.

**Comment by ivanpauno** [comment]
> I'm inclined to move forward, given that everything is green.

Agreed, I think it's ready to go

**Pull request merged**
